boards: opta: enable ADC, RS485 and ethernet reorganization

fogzot opened the pull request:
This set of changes enables the RS485 hardware connected to usart3 and provides some overlays that allow for easily running the modbus rtu_client and rtu_server samples on Opta.

A valid device tree configuration is provided for the ADCs of the 8 input channels and the sample adc_dt works out of the box. Obviously this is not the only possible configuration but it provides a good template for further customization without the need to lookup the ADC GPIOs and connections in the schematics.

Also, the `status = "okay"` was removed from ethernet configuration in the device tree to avoid always committing 16KB of RAM for the network buffers. As shown in the new overlay for the telnet sample it is sufficient to add it back to get a fully working network connection.

jukkar commented:
As this PR seems to be having three components that are being changed/reorganized, I suggest separating them to three commits for easier review and maintenance.

fogzot commented:
> As this PR seems to be having three components that are being changed/reorganized, I suggest separating them to three commits for easier review and maintenance.

I separated the changes in three commits but now I have a doubt: do you mean a single pull request with 3 commits or 3 different pull requests?

jukkar commented:
> I separated the changes in three commits but now I have a doubt: do you mean a single pull request with 3 commits

Yes, this is what I meant. No need to have three PRs.

fogzot commented:
Done. The chnages have been split into 3 commits:
1) ethernet reorganization to avoid consuming RAM when not needed
2) RS485 support
3) ADC support

fogzot committed:
boards: opta: RS485 support

This set of changes enables the RS485 hardware connected to usart3 and
provides some overlays that allow for easily running the modbus
rtu_client and rtu_server samples on Opta.

Signed-off-by: Federico Di Gregorio <fog@dndg.it>

fogzot committed:
boards: opta: ADC support

A valid device tree configuration is provided for the ADCs of the 8 input
channels and the sample adc_dt works out of the box. Obviously this is
not the only possible configuration but it provides a good template for
further customization without the need to lookup the ADC GPIOs and
connections in the schematics.

Signed-off-by: Federico Di Gregorio <fog@dndg.it>

pillo79 reviewed (commented):
Thanks for the split, looks a lot cleaner now and mostly OK to me. :+1: 
However, the Ethernet commit still contains unrelated DTS file changes (SDRAM, Flash, UART, USB). I think a separate commit for these with a little background on the changes would be helpful.

fogzot committed:
boards: opta: ethernet reorganization

This set of changes reorganize the ethernet configuration by removing the
use a regulator to enable the PHY: the correct GPIO pin is set in code
only if the network has been configured via CONFIG_NET_L2_ETHERNET.

Signed-off-by: Federico Di Gregorio <fog@dndg.it>

fogzot committed:
boards: opta: device tree cleanup

Some changes to cleanup and clarify some device tree nodes:

* removed wrong sdram2 definition
* added all internal flash slots accessible from M4
* added all internal flash slots accessible from M7
* removed CONFIG_UART_LINE_CTRL because not needed by USB CDC ACM

Signed-off-by: Federico Di Gregorio <fog@dndg.it>

fogzot commented:
> However, the Ethernet commit still contains unrelated DTS file changes (SDRAM, Flash, UART, USB). I think a separate commit for these with a little background on the changes would be helpful.

@pillo79 done. Note that the USB change in the first commit (ethernet) does not change anything: it's just a comment clarifying that the USB is assigned to M7 by default to have a working console in the common case.

pillo79 reviewed (dismissed):
Thanks, a lot cleaner now. LGTM!

facchinm commented on boards/arduino/opta/arduino_opta-common.dtsi:
```diff
@@ -58,6 +58,14 @@
 		};
 	};
 
+	sdram2: sdram@d0000000 {
```
Unluckily, Opta doesn't have 32MB of SDRAM :sweat_smile: 
Please remove the entry.

fogzot replied:
My fault, I was too fast with a copy&paste when I setup the eth part and then never asked myself why that entry was there. Removed.

pillo79 commented on boards/arduino/opta/arduino_opta_stm32h747xx_m7.dts:
```diff
@@ -23,22 +23,6 @@
-		status = "okay";
-	};
-
-	sdram2: sdram@d0000000 {
-		compatible = "zephyr,memory-region", "mmio-sram";
-		device_type = "memory";
-		reg = <0xd0000000 DT_SIZE_M(32)>;
-		zephyr,memory-region = "SDRAM2";
-		zephyr,memory-attr = <( DT_MEM_ARM(ATTR_MPU_RAM) )>;
-	};
```
This part of the change goes to the "device tree cleanup" commit. Also update that commit's description as it still refers to the SDRAM move.

fogzot replied:
Sorry for the rebase, other projects I work with have the exact opposite policy (always rebase).

Anyway, I think now everything is as required.

pillo79 reviewed (commented):
Seriously nitpicking here, but since it is so clean now it's easy to spot the issues :slightly_smiling_face: . Otherwise LGTM!

Note: please do not rebase a PR unless it is required to fix merge issues (Github will warn you if there are any!). Rebasing makes life more difficult for reviewers as the Github "Compare" button becomes useless.

pillo79 reviewed (dismissed):
Thanks for bearing with us!

fogzot commented:
@facchinm @pillo79 I tried running the failing test on my local machine in every possible way and it never fails. Do you have any suggestions about how to replicate the CI environment to understand what's happening?

pillo79 commented:
@fogzot I was always lucky that those issues usually fixed themselves with a rebase on the latest main, but it wasn't the case here. :slightly_frowning_face: 
You can download the docker image used by CI with [these instructions](https://github.com/zephyrproject-rtos/docker-image), so you could try running the test there. More info on the actual full CI setup should be in the `.github/` folder in the repo, but not sure how much can be replicated locally.

fogzot commented on boards/arduino/opta/arduino_opta_stm32h747xx_m7.dts:
```diff
@@ -124,13 +112,12 @@ zephyr_udc0: &usbotg_fs {
 		&eth_txd0_pg13
 	>;
 	pinctrl-names = "default";
-	status = "okay";
```
A third option is to add an overlay to the test to enable the ethernet. Is this acceptable?

pillo79 replied:
I vote for a working Ethernet by default. No other board has an overlay on that test, it's a hint I guess :slightly_smiling_face: 
Go ahead with the rebase.

fogzot replied:
Done. Also rephrased a bit the commit message.

pillo79 reviewed (approved):
At last! :partying_face: Thanks for this work!
